tests/test_hll.erl

```diff
@@ -0,0 +1,179 @@
+    Nodes = rt:build_cluster(4, ?CONFIG),
+
+    NodeRand1 = rt:select_random(Nodes),
+    NodeRand2 = rt:select_random(Nodes),
+
+    %% Create PB connection.
+    PBC = rt:pbc(NodeRand1),
+    riakc_pb_socket:set_options(PBC, [queue_if_disconnected]),
+
+    %% Create HTTPC connection.
+    HttpC = rt:httpc(NodeRand2),
+
+    rt:create_and_activate_bucket_type(NodeRand1,
+                                       ?HLL_TYPE1,
+                                       [{datatype, hll},
+                                        {hll_precision, ?SET_P}]),
+
+    rt:create_and_activate_bucket_type(NodeRand1,
+                                       ?HLL_TYPE2,
+                                       [{datatype, hll}]),
+
+    pb_tests(PBC, riakc_pb_socket, ?BUCKET1, Nodes),
+    http_tests(HttpC, rhc, ?BUCKET2, Nodes),
+
+    %% Stop PB connection.
+    riakc_pb_socket:stop(PBC),
+
+    %% Clean cluster
+    rt:clean_cluster(Nodes),
+
+    pass.
+
+http_tests(C, CMod, Bucket, Nodes) ->
+    lager:info("HTTP CLI TESTS: Create new Hll DT"),
+    S0 = riakc_hll:new(),
+
+    add_tests(C, CMod, Bucket, S0),
+
+    HllSet0 = get_hll(C, CMod, Bucket),
+    check_precision_and_reduce_test(C, CMod, Bucket, ?DEFAULT_P, HllSet0),
+
+    partition_write_heal(C, CMod, Bucket, Nodes),
+
+    HllSet1 = get_hll(C, CMod, Bucket),
+    check_precision_and_reduce_invalid_test(C, CMod, Bucket, ?DEFAULT_P - 1,
+                                       HllSet1),
+
+    ok.
+
+pb_tests(C, CMod, Bucket, Nodes) ->
+    lager:info("PB CLI TESTS: Create new Hll DT"),
+    S0 = riakc_hll:new(),
+
+    add_tests(C, CMod, Bucket, S0),
+
+    HllSet0 = get_hll(C, CMod, Bucket),
+    check_precision_and_reduce_test(C, CMod, Bucket, ?SET_P, HllSet0),
+
+    partition_write_heal(C, CMod, Bucket, Nodes),
+
+    HllSet1 = get_hll(C, CMod, Bucket),
+    check_precision_and_reduce_invalid_test(C, CMod, Bucket, ?SET_P - 1, HllSet1),
+
+    ok.
+
+add_tests(C, CMod, Bucket, NewS) ->
+    add_element(C, CMod, Bucket, NewS, <<"OH">>),
+    {ok, S1} = CMod:fetch_type(C, Bucket, ?KEY),
+    ?assertEqual(riakc_hll:value(S1), 1),
+
+    add_elements(C, CMod, Bucket, S1, [<<"C">>, <<"A">>, <<"P">>]),
+    {ok, S2} = CMod:fetch_type(C, Bucket, ?KEY),
+    ?assertEqual(riakc_hll:value(S2), 4),
+
+    add_redundant_element(C, CMod, Bucket, S2, <<"OH">>),
+    {ok, S3} = CMod:fetch_type(C, Bucket, ?KEY),
+    ?assertEqual(riakc_hll:value(S3), 4).
+
+partition_write_heal(C, CMod, Bucket, Nodes) ->
+    lager:info("Partition cluster in two to force merge."),
+    [N1, N2, N3, N4] = Nodes,
+    PartInfo = rt:partition([N1, N2], [N3, N4]),
+
+    lager:info("Write to one side of the partition"),
+    {ok, S0} = CMod:fetch_type(C, Bucket, ?KEY),
+    add_element(C, CMod, Bucket, S0, <<"OH hello there">>),
+    {ok, S1} = CMod:fetch_type(C, Bucket, ?KEY),
+    ?assertEqual(riakc_hll:value(S1), 5),
+
+    lager:info("Heal"),
+    ok = rt:heal(PartInfo),
+    ok = rt:wait_for_cluster_service(Nodes, riak_kv).
```

Comment:
Shouldn't you then test that the new value is present after healing? Or shouldn't you have "incremented" on both sides and checked that the total was correct. I'm not clear on what is being tested here.